src/lib/transition_frontier/transition_frontier.ml

```diff
@@ -391,24 +472,64 @@ struct
         (* 2.b *)
         if distance_to_parent > max_length t then (
           (* 2.b.I *)
-          let new_root_hash = List.hd_exn (hash_path t node.breadcrumb) in
+          let heir_hash = List.hd_exn (hash_path t node.breadcrumb) in
+          let heir_node = Hashtbl.find_exn t.table heir_hash in
           (* 2.b.II *)
-          let garbage_immediate_successors =
-            List.filter root_node.successor_hashes ~f:(fun succ_hash ->
-                not (State_hash.equal succ_hash new_root_hash) )
+          let heir_heirs_nodes =
+            List.map heir_node.successor_hashes ~f:(Hashtbl.find_exn t.table)
           in
           (* 2.b.III *)
+          let bad_hashes =
+            List.filter root_node.successor_hashes
+              ~f:(Fn.compose not (State_hash.equal heir_hash))
+          in
+          let bad_nodes = List.map bad_hashes ~f:(Hashtbl.find_exn t.table) in
+          (* 2.b.IV *)
           let garbage =
-            t.root
-            :: List.bind garbage_immediate_successors
-                 ~f:(successor_hashes_rec t)
+            (t.root :: heir_hash :: heir_node.successor_hashes)
+            @ List.bind bad_hashes ~f:(successor_hashes_rec t)
           in
-          t.root <- new_root_hash ;
           List.iter garbage ~f:(Hashtbl.remove t.table) ;
-          (* 2.b.IV *)
-          Ledger.Mask.Attached.commit
-            (Inputs.Staged_ledger.ledger
-               (Breadcrumb.staged_ledger root_node.breadcrumb)) ) ;
+          (* 2.b.V *)
+          let new_root_crumb, new_heir_heirs_crumb, proof_txns =
+            Breadcrumb.commit ~root:root_node.breadcrumb
+              ~heir:heir_node.breadcrumb
+              ~bads:(List.map bad_nodes ~f:breadcrumb_of_node)
+              ~heir_heirs:(List.map heir_heirs_nodes ~f:breadcrumb_of_node)
+          in
+          (* 2.b.VI *)
+          let () =
+            match proof_txns with
+            | Some txns ->
+                (* TODO: do these in batch! *)
```

Comment:
You should probably do that now, not later. If you want to merge this anyway, please make an issue because this is very important with the way mask's work.